### ISSUE TITLE
Brain dump of istanbul - nyc deltas

### ISSUE THREAD

**Issue by gotwarlost** [istanbuljs/nyc#309]
Differences and gaps in `istanbul` versus `nyc` as a brainstorm doc. We should create individual issues for these in a second pass and tag them as `now` versus `later` for the ones for which something needs to be done.

In general, the highest priority is for gaps that affect node.js projects. Browser use-cases are second class citizens for the first release.
## istanbul configuration deltas

Type `istanbul help config` for full list. We posit a migration script that can turn a `istanbul.yaml` file to an `nyc` config.
### instrumentation (instrument and cover commands)
1. instrumentation root: root below which to instrument files (default: process.cwd, I think this is _rarely_ used)
2. default-excludes [bool]: exclude `node_modules/*`, `* test/**`, `tests/**` (default: true, false means consider everything)
3. compact [bool] - emit compact code (default: true).
4. preserve-comments [bool] whether to preserve comments (default: false)
5. save-baseline [bool] - whether the `istanbul instrument` command should save a baseline coverage file
#### Conclusions

default-excludes MUST be solved. This is a BIG diff for people migrating from istanbul to nyc. The simplest approach is to let the migration script handle it by adding appropriate -x options in the nyc config.  This doesn't pollute the nyc codebase and provides a backwards-compatible migration path.

No need to allow files from under `node_modules` to be instrumented for now (although people in istanbul-land have specifically wanted this).

(1), (3) and (4) can be opinions in nyc for now until we find valid use-cases for exposing these options.

(5) may be an issue for people wanting a baseline for instrumentation done for browser tests (since it is basically the  `--all` implementation for browser testing). We can let people complain before we add it. The browser use-case needs some thinking anyway and we can plod along with old versions of istanbul until we close the gaps.
### hooks
- [x] handle-sigint (nyc does this by default?) **this is handled**
- post-require-hook (do NOT support in nyc. This is some legacy garbage for supporting YUI's loader)
#### Conclusions

I don't think any changes are needed here as long as signals are handled by nyc out of the box.
### reporting
1. summarizer [pkg|flat|nested] - summarizer to use for reporting (new feature in 1.0, default: pkg)
2. watermarks: { statements: [50, 80], lines: [50, 80], branches: [50, 80], lines: [50, 80]}
   - low and high watermarks for reporting low|medium|high coverage. Affects report colorization etc.
3. report-config: object containing properties keyed by report name (detailed config for each report)
#### Conclusions
- In general, **reporting has the biggest delta** and we may need a report config file because the crap above like watermarks and individual report config is hard to do with command line flags. 
- IMO, lack of watermarks and report config is a biggish delta for people that use these features and we should try and cover this gap. 
- Summarizer should be exposed on the command line (although, this is not advertised even for istanbul 1.0 so we can punt on it for the initial release).
### check-coverage
- "each" options  to check coverage on every file individually
- "each" config has an exclusions array that omits the check for specific files/ globs
- "global" config also has an exclusions array that eliminates matching files before calculating total statements, branches, lines, functions
#### Conclusions
- Do nothing for the first release. Figure out how to introduce this cleanly in nyc later.
## istanbul instrument command
- nyc's interface is better in general. 
- nyc does not support baseline json
- istanbul has a -compelete-copy flag that copies all non-JS files to the output tree as well
### Conclusions
- Nothing needed. Not a blocker for first release. Browser use-case has not been thought through fully.
### istanbul report command
- supports input root directory to look for coverage files (the --temp-directory option added in the latest PR although it is not fully implemented for other commands)
- nyc has no support for custom reporting config
#### Conclusions

Plug the second gap if possible.
## New features and API stability

Once the 1.0 API is locked down we won't be able to make backwards incompatible changes to it. We might as well think through some of the issues here even if we consciously
decide to not do anything about it.
### Coverage JSON format
- should we have hashing/ signature generation for the shape of a coverage object (statement, branch, function maps) in the file coverage object itself
  (implemented by `istanbul-lib-coverage`)?  Currently the babel plugin implements sigs, nyc itself does some hashing and so on. Having an inbuilt sig deep 
  in the file coverage object will allow merges to fail fast when the coverage objects for the same file have different signatures.
- The coverage map is now just a map keyed by file paths. Should we have an explicit `files` key under which everything is stored?
  This will allow us to detect non-coverage maps loaded from JSON _and_ it will allow us to stash additional metadata in a `meta` key or something.
  As long as istanbul-lib-coverage is the only lib that processes coverage objects (which may not be the case in nyc today), we can make the processing seamless for nyc.

Experience suggests that people _will_ poke into the internals of the coverage format and taking the high road to tell people it is a private interface
will not work. Thus, any changes here must be done prior to API freeze.
## Process gaps
1. Need to finalize how issues will be reported across the thousand libraries and tools we have and provide users with a brain-dead mechanism for this.
2. Said process must specifically have an end-of-life plan for issues reported on gotwarlost/istanbul, even if it is as simple as "close everything
   and ask users to open an issue in <the correct place> if it sill affects you". We are _not_ carrying this over for sanity.
3. We need to have some guidance on how issues will be reported. Perhaps you guys already have this in place. Specifically, we need people to give us
   basic information about library versions etc. This can leak into "technical" so that people can easily provide this info with one command.
   One of the biggest problems I've had is that issues start with "I have a grunt/ gulp/ build-system-du-jour task that doesn't work" and I don't
   know _anything_ about said build-system-du-jour. We need to help people to report issues in a way we can attack them.
4. Write migration script (from istanbul.yaml to nyc config) and package as part of `nyc` (or `istanbul`?)?
5. Add API docs for all APIs. I have started this work but I don't yet have a process for publishing the docs on gh-pages etc.
6. Write a HOWTO for "istanbul to nyc"
7. Formally deprecate istanbul and advertise this using social media. I'm the most anti-social person I know so someone else has to do this :)
8. Follow up with the babel guys on generator perf improvements and source maps fixes
9. Sample repos: Come up with common use cases, have a sample repo showing how to get coverage for that use-case. Add to list as necessary. Examples:
   - ES5 code
   - Babel
   - Typescript
   - Test harnesses that need specific config
   - Browser tests (later)
10. update current sample repos with state of the art.
11. (advanced) exercise the sample repos as part of testing every release to ensure no regressions. Allow people to add sample repos.
12. Figure out what we are going to do with browser use-cases.
## Internal
- Add tests for istanbul-reports
- Add tests for istanbul-lib-source-maps.

Right now PRs are a nightmare for these repos because of lack of tests. Should we consider this a blocker for first release?

**Comment by gotwarlost** [comment]
True story. I was jet-lagged and decided to come to work early at 6:00 to work on "day job" stuff. Badge disabled after month-long vacation. No internet access. So had the chance to bang this out between 6:00 and 8:00 with no interruptions :)

**Comment by bcoe** [comment]
@gotwarlost awesome, this gives us a great place to start working from :) let's gradually turn these into check boxes.

**Comment by gotwarlost** [comment]
@bcoe - this is just a brain dump. Don't let the length of the list overwhelm you :)

My 2c, these are the must haves:

* Migration script from istanbul.yaml to nyc
* Some report config mechanism including config for watermarks (not a show stopper for release but we should start discussion now)
* Bug process
* Decisions on coverage JSON changes, if we want to make them

After that, we could maybe invite brave folks to beta test migration from istanbul to nyc and see what the feedback is for further prioritization.
 
/cc @davglass - Dav, if you have some time, could you migrate some of your repos to use `nyc` instead of `istanbul` and let us know what the pain points were?

**Comment by bcoe** [comment]
@gotwarlost wonderful writeup. would love some help splitting this into tasks, and getting over the finish line (CC: @novemberborn, @jamestalmage, @JaKXz, @kentcdodds).

I'll definitely start digging in this weekend too, my kingdom for more OSS hacking time!

**Comment by jamestalmage** [comment]
> should we have hashing/ signature generation for the shape of a coverage object

I think this is a good idea. Would a hash of the coverage shape be too permissive? Why not just a combined hash of the transform input and output?

```js
const outputCode = transform(inputCode);

const hash = md5Hex([inputCode, outputCode]);
```

Maybe include both? Fail if shape of the coverage object is different (unmergeable), warn if stricter hash is different (`File XXX appears to have been instrumented using a different version / settings`)?

**Comment by bcoe** [comment]
@istanbuljs/nyc I've created a milestone for moving towards an Istanbul 1.0 release (as it relates to nyc):

https://github.com/istanbuljs/nyc/milestone/2

@gotwarlost feel free to add additional issues here, and stack rank them as you see fit; others should feel free to grab any off the stack that they're interested in working on.

## tldr;

A few of the issues listed in this brain dump, largely relate to documentation:

* having appropriate code samples.
* documenting the migration process.

I've created [istanbuljs.github.io](https://github.com/istanbuljs/istanbuljs.github.io/issues/1), we should start pulling together a community website for addressing some of these requirements.

**Comment by eberhara** [comment]
@gotwarlost @bcoe any news on these topics?
- "each" options to check coverage on every file individually
- set low and high watermarks for reporting low|medium|high coverage

thanks!

**Comment by bcoe** [comment]
@eberhara we're finally landing configuration for low and high watermarks:

https://github.com/istanbuljs/nyc/pull/493

No implementation of each yet; would love a helping hand if anyone wants to run with it.

**Comment by JaKXz** [comment]
@gotwarlost would you mind making this into a GH project board? I think it'd be easier to visualize progress there. :)

**Comment by bcoe** [comment]
I think we've knocked most of these out of the park.
